feat: implement multi-output support with scaling

bigFin opened the pull request:
## Description
Papertoy but on multiple displays
- independent configuration of frame rate, resolution, and scaling
- handles display rotation
- tested on Niri and Hyprland, both Nixos and Arch

- **Per-Output `Paper` Instances:** Manage each output independently. A new `Paper` struct encapsulates all state necessary for rendering on a single output (including `WlrSurface`, `Shader`, and OpenGL contexts).  Subsequently decided to try a shared OpenGL context to see if there were any per benefits, if so it wasn't major.

- **Respect Compositor Configuration:** The application now correctly processes `zwlr_layer_surface_v1.configure` events, which provide the authoritative logical dimensions from the compositor. This ensures the application always resizes correctly based on compositor instructions.

- **Output Orientation Support:** The application now handles `wl_output.geometry` events, which communicate display rotation. It swaps width and height when a display is rotated by 90 or 270 degrees, tested on 3440x1440 vs 1440x3440.

- **`wlr_layer_surface` Anchoring:** Layer surfaces are now anchored to all four edges (`top`, `bottom`, `left`, `right`). This explicitly tells the compositor to stretch the wallpaper to fill the entire output area- this was needed for Niri when the surface might not cover the whole display when in vertical orientation. 

- **`display scaling`:**  Added a scale parameter to uniformly adjust the `resolution`. 

## How to Test
**1. Performance Mode (Recommended for 4K)**
Render at half resolution (1080p on 4K) to save GPU power, but keep the window fullscreen.
```console
$ papertoy shader.glsl --output "id=HDMI-A-1,scale=0.5"
```

**2. Multi-Monitor Setup**
Configure a high-refresh main monitor and a slower secondary monitor.
```console
$ papertoy shader.glsl \
    --output "id=DP-1,frame-rate=144" \
    --output "id=HDMI-A-1,scale=0.5,frame-rate=30"
```

**3. Custom Resolution**
Force a specific aspect ratio or size.
```console
$ papertoy shader.glsl --output "id=DP-1,resolution=800x600"
```

**4. All displays at native settings**
Automatically render papertoy on all displays
```console
$ papertoy shader.glsl 
```

Verify that animations are smooth on all displays and that the wallpaper correctly fills the entire screen area, including any fractionally scaled or rotated displays.

sin-ack commented on src/main.zig:
```diff
@@ -731,66 +811,142 @@ pub fn main() !u8 {
         }
-        target_frame_rate = @intCast(frame_rate);
+        custom_frame_rate = @intCast(frame_rate);
     }
```
It should be possible to set the frame rate on a per-output basis. The user may want the animation on two monitors, but for their second monitor may choose to set a low framerate to conserve resources, for example.

The command-line syntax I'm thinking of is something like: `--output "id=DP-1,resolution=1280x720,frame-rate=30"`. `id` should be required but anything else should be optional and default to whatever's native, just like with the single-output variant.

sin-ack commented on src/main.zig:
```diff
@@ -266,6 +270,14 @@ const WlrSurface = struct {
     /// The wlroots surface.
     wlr_surface: *zwlr.LayerSurfaceV1,
+    
+    /// Pending configuration from the compositor.
+    pending_config: ?struct {
+        serial: u32,
+        width: u32,
+        height: u32,
+    } = null,
+    config_dirty: bool = false,
```
I like this change, but it should be its own commit, not squashed to a single commit. Please see the repo's commit history for examples on how fixes are committed.

bigFin replied:
Got it, thank you for the guidance and kindly pardon the slop.

sin-ack commented on src/main.zig:
```diff
@@ -319,11 +331,16 @@ const WlrSurface = struct {
 
         // We want to ignore any exclusive zones set by other surfaces.
         self.wlr_surface.setExclusiveZone(-1);
-        self.wlr_surface.setSize(self.width, self.height);
-        self.wl_surface.setBufferScale(@intCast(self.scale));
+        self.wlr_surface.setSize(self.destination_width, self.destination_height);
+        
+        if (fractional_scale_manager) |_| {
+            self.wl_surface.setBufferScale(1);
+        } else {
+            self.wl_surface.setBufferScale(@intCast(self.scale));
+        }
 
-        // TODO: Make the user set this.
-        self.wlr_surface.setAnchor(.{ .top = true, .left = true });
+        // Anchor to all 4 sides to fill the screen.
+        self.wlr_surface.setAnchor(.{ .top = true, .bottom = true, .left = true, .right = true });
```
When I tried this with Hyprland I had issues. Could you verify that current Hyprland works fine with this?

bigFin replied:
Just tested, it works on hyprland with no issues for me.

sin-ack commented on src/main.zig:
```diff
@@ -384,51 +401,91 @@ const WlrSurface = struct {
+        if (logical_width == 0 or logical_height == 0) {
+            logical_width = self.output.width;
+            logical_height = self.output.height;
+        }
```
What is a case where a new configuration sends a 0x0 size? Did you ever encounter this in the wild? I'm testing with Niri and even though it's not a "conventional compositor" so to speak, it still sends the expected output size.

Also, please split this into its own commit.

bigFin replied:
This is informed by [wlr-layer-shell-unstable-v1.xml](https://github.com/swaywm/wlr-protocols/blob/master/unstable/wlr-layer-shell-unstable-v1.xml) " If the width or height arguments are zero, it means the client should decide its own window dimension". Not something I saw in the wild.

sin-ack commented on src/main.zig:
```diff
@@ -384,51 +401,91 @@ const WlrSurface = struct {
+        // Custom resolution overrides everything.
         if (self.custom_resolution) |resolution| {
-            expected_width = resolution.width;
-            expected_height = resolution.height;
+            logical_width = resolution.width;
+            logical_height = resolution.height;
         }
```
The 0x0 check above can be added as an else-if branch to this.

bigFin replied:
If the 0x0 check was an else-if to the custom resolution check and both a custom resolution was set and the compositor sent 0x0, this would override the custom resolution. I think we wanna process the 0x0 first then let custom override take precedence.

sin-ack commented on src/main.zig:
```diff
@@ -319,11 +331,16 @@ const WlrSurface = struct {
-        self.wl_surface.setBufferScale(@intCast(self.scale));
+        self.wlr_surface.setSize(self.destination_width, self.destination_height);
+        
+        if (fractional_scale_manager) |_| {
+            self.wl_surface.setBufferScale(1);
+        } else {
+            self.wl_surface.setBufferScale(@intCast(self.scale));
+        }
```
Make sure to rebase on top of master now that #28 is merged.

sin-ack commented on src/main.zig:
```diff
@@ -605,6 +666,61 @@ fn handleArgsError(err: zig_args.Error) !void {
+        errdefer {
+            self.global_attributes.deinit();
+            self.shader.destroy(allocator);
+        }
```
Move the `global_attributes.deinit()` under the `init` call for consistency.

sin-ack commented on src/main.zig:
```diff
@@ -731,66 +811,142 @@ pub fn main() !u8 {
+                });
+            }
+            return 1;
+        }
```
This option should be additive and parsed as described above. We can keep the default as "all outputs, all native".

sin-ack commented on src/main.zig:
```diff
@@ -605,6 +666,61 @@ fn handleArgsError(err: zig_args.Error) !void {
     try printUsage();
 }
 
+const Paper = struct {
```
I like the name `Paper` :)

sin-ack commented on src/main.zig:
```diff
@@ -605,6 +666,61 @@ fn handleArgsError(err: zig_args.Error) !void {
+    global_attributes: GlobalAttributes,
+    render_frame: bool = false,
+    next_frame_time: u64 = 0,
+    expected_frame_time_ns: u64 = 0,
```
Please document these parameters.

sin-ack commented on src/main.zig:
```diff
@@ -731,66 +811,142 @@ pub fn main() !u8 {
+        for (papers.items) |paper| {
+            // Check if we should render this paper
+            if (paper.shader.frame > 0) {
+                if (using_custom_frame_rate) {
```
As part of making the frame-rate setting per-output, since the fields used for custom FPS rendering and callback-based rendering are mutually exclusive, we can group those fields with a tagged union like `NextFrameStrategy` and `switch` on that to do the right thing.

sin-ack commented on src/main.zig:
```diff
@@ -731,66 +811,142 @@ pub fn main() !u8 {
+            if (rendered_count == 0) {
+                if (display.dispatch() != .SUCCESS) return error.DispatchFailed;
+            }
+        }
```
With mixed next-frame strategies we'll want to use [`display.dispatchTimeout`](https://wayland.freedesktop.org/docs/html/apb.html#Client-wayland-client-core_8h_1ad4ce9d6fb41cf2b36d34a4626ef36561) so that we can wait for either any frame callback or until one of the frame timings expire, whichever comes first.

bigFin replied:
How about this- wl.Display in wayland.zig does not expose dispatchTimeout so instead poll waits for events or the `smallest_custom_sleep_time' based on at least one custom fps across the papers.

sin-ack reviewed (changes requested):
Thanks for the contribution. From what I can tell you're getting help from an LLM for this change. That's fine, but:

- The PR description and commit message seem to be generated too. Please describe the change with your own words. LLMs are good at explaining *what* happened but they're not good at explaining *why*.
- Unfortunately LLM providers really like to RL their coding models into doing a single commit for all changes, resulting in a hard-to-review PR because it's more difficult to tell which line of code belongs to which change. Please use [atomic commits](https://www.aleksandrhovhannisyan.com/blog/atomic-git-commits/) for each separate change being made. I recommend you to look at this repo's history for how I expect things to be structured. You can ask the LLM about how to modify Git history, or make it do the split by itself (although I wouldn't really trust an LLM to run commands on my behalf, `git reflog` is your friend here).
- As I detailed below, one of the reasons I kinda postponed working on this was due to not being able to figure out how to handle multiple frame-rate options within the same loop (I had initially planned to spawn multiple threads and run a queue on each). The changes below seem to mostly address that nicely, but they use the same frame-rate for all outputs. Instead, make the `--output` flag configure individual outputs which you can then use as the basis for creating `Paper` objects.
- Since as part of the changes the command-line options will change, please make sure to update the README too.

bigFin committed:
fix: correct fractional scaling calculation

bigFin committed:
feat(surface): introduce pending config and event-driven size updates

bigFin committed:
feat(multi-output): implement multi-output rendering

bigFin committed:
feat(config): per-output configuration and frame rate strategy

bigFin committed:
docs: update README with new --output flag usage

bigFin committed:
docs: add documentation for Paper struct fields

bigFin commented:
TYVM for the review and many thanks for taking the time to help me clean up the slop. I divided the chungus squash commit to be atomic. Independent configurable frame rates works. I tested this on Niri and Hyprland and both work for me.

bigFin committed:
fix(output): Correct memory leak in Output name handling

bigFin committed:
perf(shader): Use glBufferSubData for efficient uniform updates

bigFin committed:
refactor(egl): Extract GLContext and update WlrSurface to use it via pointer

bigFin committed:
perf(core): Implement shared OpenGL context for multi-output rendering

bigFin committed:
feat(perf): Add render scaling and fix shared context viewport

bigFin committed:
docs: Update README with render scaling and configuration examples

bigFin commented:
I tried a few things help the user manage gpu utilization
- Add a render scaling parameter so the user can choose to render at a lower res. Some combination of this and lower fps can significantly reduce gpu use.
- shared OpenGL context for multi output rendering. Im not sure if this really makes a big impact in practice.

bigFin committed:
fix: add initial roundtrip to prevent startup crash on Arch Wayland

bigFin committed:
flake: add pure nixGLIntel wrapper, drop impure auto-detect

- Integrates `nixgl` via overlay to ensure pure evaluation.
- Adds `nixGLIntel` wrapper for x86 Linux, enabling execution on non-NixOS distros (like Arch).
- Drops the generic `nixGL` auto-detect wrapper as it relies on `builtins.currentTime`, which is impure and fails `nix flake check` in CI.
- Updates `flake.lock` with `nixgl` input.

sin-ack commented:
Sorry for the lack of a review, I was on a trip. I'll take a closer look at this on Sunday.

> Add a render scaling parameter so the user can choose to render at a lower res.

The `--resolution` argument already does this. The result will be scaled to the output's size but the surface size is set to the argument's value so that rendering at a smaller resolution is possible.

> shared OpenGL context for multi output rendering.

Some benchmarks for render time would be cool!

bigFin commented:
Hope you had a good trip! No rush. As you can see the scope of this PR has extended a bit over the week as I tinkered and tried a few ideas, which has been fun and a good learning experience. Thanks for the feedback and for working with me on this! 

Good point about the `resolution` argument, this type of optimisation was always possible. The uniform `scale` is just a convenience.

Benchmarks on the shared OpenGL context was interesting technically but not so impactful in my case and probably for most desktop users. Sharing the context reduces the cost of switching surfaces on the driver side ~5% reduction from`eglMakeCurrent` (-2.4 us). This shifts the work to the render to swap the viewport +8% (+1.5 us). So maybe we save 1 us per frame with the shared context on my system, seems negligible to me but maybe it adds up to something when using this to drive a lot of outputs. 

**Commit:** `58d7610` (Shared Context + Telemetry) vs Reverted `HEAD` (Independent Contexts + Telemetry)
**Methodology:**
- **Control:** Independent `GLContext` per Output.
- **Experiment:** Single Shared `GLContext` across all Outputs.
- **Metric:** Internal Telemetry (average nanoseconds per frame) + `perf stat`.
- **Workload:** `shaders/stochastic-asym-quads.glsl` on 2 outputs (4K@119Hz, UWQHD@100Hz).
- **Duration:** Fixed 600 frames per run (approx 3 seconds).
## Results (Average of 3 runs, Normalized per Render)

| Metric | Independent Contexts (Control) | Shared Context (Experiment) | Diff (ns) | % Diff |
| :--- | :--- | :--- | :--- | :--- |
| **Total System CPU Time (ms)** | ~704 | ~713 | **+9** | **+1.3%** |
| **MakeCurrent (Context Switch)** | ~47,978 | ~45,601 | **-2,377** | **-5.0%** |
| **Render (GPU Submission)** | ~17,517 | ~18,985 | **+1,468** | **+8.4%** |
| **SwapBuffers (Present)** | ~1,780,501 | ~1,563,170 | -217,331 | -12.2% |
| **OS Context Switches (total)** | ~8,790 | ~9,007 | +217 | +2.5% |


edit---- 
ran the perf test a bit longer and also added GPU metrics- variance was whack on the short runs now its <1% across all metrics. Lower VRAM use is also nice.
- **Duration:** Fixed 12000 frames per run (approx 60 seconds) for maximum stability.

## Results (Average of 3 runs, Normalized per Render)

| Metric | Independent Contexts (Control) | Shared Context (Experiment) | Diff | % Diff |
| :--- | :--- | :--- | :--- | :--- |
| **VRAM Usage (Delta)** | **405 MiB** | **333 MiB** | **-72 MiB** | **-17.8%** |
| **Avg MakeCurrent (ns)** | ~44,943 | ~42,587 | -2,356 | **-5.2%** |
| **Avg Render (ns)** | ~15,983 | ~15,709 | -274 | **-1.7%** |
| **Avg SwapBuffers (ns)** | ~2,649,089 | ~2,841,274 | +192,185 | +7.3% |
| **Total System CPU Time (ms)** | ~9,431 | ~9,305 | -126 | **-1.3%** |


Seems like a win to me

bigFin committed:
perf(shader): update uniforms with bufferSubData

bigFin committed:
feat: add multi-output rendering

bigFin committed:
docs: document per-output configuration

bigFin committed:
Respect frame callbacks when capping frame rate

bigFin committed:
nix: update zig-wayland scanner fix
